rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
+    } else {
+      ddt = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor_prim(DDS_DYNAMIC_SEQUENCE, member->name_, 1, (uint32_t*)(&member->array_size_), type)); 
+    }
+  } else {
+    ddt = dds_dynamic_type_create(dds_ppant,
+       get_dynamic_type_descriptor_prim(DDS_DYNAMIC_SEQUENCE, member->name_, 0, nullptr, type)); 
+  }
+
+  dds_return_t ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(ddt, member->name_));
+  if (ret != DDS_RETCODE_OK)
+      RCUTILS_LOG_ERROR("failed to add type member sequence/array with prim type");
+}
+
+template<typename MemberType>
+static void dynamic_type_add_array(
+  dds_dynamic_type * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_t ddt)
+{
+  dds_dynamic_type_t dseq;
+
+  if (member->array_size_) 
+  { 
+    if (!member->is_upper_bound_)
+    {
+      dseq = dds_dynamic_type_create(dds_ppant, 
+       get_dynamic_type_descriptor(DDS_DYNAMIC_ARRAY, member->name_, 1, (uint32_t*)(&member->array_size_), ddt));
+    } else {
+      dseq = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor(DDS_DYNAMIC_SEQUENCE, member->name_, 1, (uint32_t*)(&member->array_size_), ddt));
+    }
+  } else {
+    dseq = dds_dynamic_type_create(dds_ppant,
+     get_dynamic_type_descriptor(DDS_DYNAMIC_SEQUENCE, member->name_, 0, nullptr, ddt)); 
+  }
+
+  dds_return_t ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor(dseq, member->name_));
+  if (ret != DDS_RETCODE_OK)
+    RCUTILS_LOG_ERROR("failed to add type member sequence/array with dynamic type");
+}
+
+template<typename MemberType>
+static void dynamic_type_add_member(
+    dds_dynamic_type_t * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_kind_t type)
+{
+  dds_return_t ret;
+  if (type == rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING ||
+      type == rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE)
+    return;
+
+  if (!member->is_array_)
+  {
+    ret = dds_dynamic_type_add_member(dstruct, get_dynamic_member_descriptor_prim(type, member->name_));
+    if (ret != DDS_RETCODE_OK)
+      RCUTILS_LOG_ERROR("failed to add type member: %d", type);
+  } else {
+    dynamic_type_add_array(dstruct, dds_ppant, member, type);
+  }
+}
+
+static void dynamic_type_register(struct sertype_rmw * st, dds_dynamic_type_t dt, dds_entity_t dds_ppant)
+{
+  dds_typeinfo_t *type_info;
+  auto rc = dds_dynamic_type_register(&dt, &type_info);
+  if (rc != DDS_RETCODE_OK)
+    RMW_SET_ERROR_MSG("dds_dynamic_type_register failed to register type");
+  
+  dds_topic_descriptor_t *desc;
+  rc = dds_create_topic_descriptor(
+    DDS_FIND_SCOPE_GLOBAL, dds_ppant, type_info, 0, &desc);
+  if (rc != DDS_RETCODE_OK)
+    RMW_SET_ERROR_MSG("dds_create_topic_descriptor failed to create descriptor");
+
+  if (st)
```

Comment:
Do we ever get here when `st == nullptr`? It seems to me there is not much reason to.

Reply:
Yes you are right this step is not needed here.